### PR TITLE
[Distributed Strategy] GLM-4.5 model accuracy alignment

### PR DESCRIPTION
**Pull request by zhanghonggeng**
### PR Category

Distributed Strategy

### PR Types

New features

### Description

GLM-4.5 模型精度对齐（PaddleFleet vs Megatron-LM），所有对齐逻辑通过 use_accuracy_compatible 开关控制，开关关闭时不影响原有代码行为。

核心改动：

moe_utils.py — 实现确定性 permute/unpermute（gather+sum fp32 PyLayer），消除 scatter_add 非确定性累加序差异

moe_router.py — gate weight dtype 改为 bf16（对齐 MG 默认行为）；router scoring 使用 re-sigmoid 计算 gates_ori；TopK 归一化使用 fp64 求和；aux_loss 计算对齐 MG 公式；take_along_axis 替换为 gather_nd（对齐反向梯度路径）

moe_layer.py — 三路 clone（ThreePathCloneAlignMG PyLayer）控制 bf16 backward 累加顺序；小 M expert padding（M<17 → 32）消除 cublasLt tiny-M kernel 差异；EP>1 时强制 alltoall dispatcher

fused_bias_swiglu.py — 对齐模式下使用 eager SwiGLU forward/backward（非 fused），避免 jit_fuser kernel 数值差异

tensor_parallel/layers.py — forward 使用 weight.T.contiguous() + paddle.matmul 对齐 GEMM layout；expert weight 的 wgrad 使用 fp32 直接累加到 main_grad

language_loss.py — loss 求和使用 fp64 精度；模拟 EP allreduce 的 bf16 累加序

model_parallel_config.py — 对齐模式下自动启用 deterministic_mode=True

rotary_pos_embedding.py — 移除已无用的 use_accuracy_compatible 分支（CPU/GPU 计算 inv_freq 数值等价）

### 是否引起精度变化

否

### CODE REVIEW

File: src/paddlefleet/transformer/moe/moe_layer.py [hunk 1]
```diff
@@ -712,20 +714,51 @@ def expert_forward(
...
-                per_token_scale, num_or_sections=tokens_per_expert, axis=0
-            )
         for i, chunk in enumerate(chunks):
             if tokens_per_expert[i] == 0:
                 continue
             chunk = chunk.contiguous()
             current_expert_idx = i + self.moe_rank * self.num_experts_per_device
             expert = self.experts[current_expert_idx]
-            if scale_chunks is None:
+            if self.use_accuracy_compatible and 0 < int(chunk.shape[0]) < 17:
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/moe/moe_layer.py]: P1 这里直接读取 `self.use_accuracy_compatible` 会让现有必跑单测在进入非空 expert 分支时触发 `AttributeError`。

`test_ai_moe_layer_2.py` 使用 `MinimalMoE` 调用 `MoELayer.expert_forward`，`test_ai_moe_layer_7.py` 也通过 `MoELayer.__new__` 构造轻量对象；两者都没有该属性，且均未列入单卡测试禁用清单。请改为 `getattr(self, "use_accuracy_compatible", False)`，或同步初始化这些受支持的测试对象，并补充 tiny-M padding 的前向/反向用例。

File: src/paddlefleet/models/common/language_loss/language_loss.py [hunk 2]
```diff
@@ -439,10 +443,31 @@ def _cast_loss_func(logits, labels):
...
-                loss = loss / lossmask.sum()
+                if self.use_accuracy_compatible:
+                    _flat = loss.cast(paddle.float32).reshape([-1]) * lossmask
+                    loss_sum = (
+                        _flat.cast(paddle.float64).sum().cast(paddle.float32)
+                    )
+                    _count = lossmask.sum()
+                    import paddle.distributed as _pdist
+
+                    _ep_group = get_expert_model_parallel_group(
```
Comment (risemeup1111) [inline src/paddlefleet/models/common/language_loss/language_loss.py]: P2 这里绕过了构造函数已保存的 `self.pg_collection.ep`，显式传入 `ProcessGroupCollection` 时可能按错误的 EP size 模拟累加。

该构造参数就是仓库支持的自定义进程组入口；若全局 `parallel_state` 中的 EP group 为 `None` 或与传入集合不同，`_ep_size` 会退化为 1 或取到另一组，bit-wise loss 路径便不再对应实际拓扑。请优先从 `self.pg_collection.ep` 获取 group，并覆盖显式进程组与 EP>1 的回归用例。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
|---|---|---|---|
| 1 | [moe_layer.py:988](https://github.com/PaddlePaddle/PaddleFleet/pull/1627#discussion_r3701792481) | P1 | ✅ |
| 2 | [language_loss.py:455](https://github.com/PaddlePaddle/PaddleFleet/pull/1627#discussion_r3701793746) | P2 | ✅ |

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/moe/moe_layer.py (hunk 1 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/moe/moe_layer.py]: 当前提交已将 `expert_forward` 改为 `getattr` 并补充 tiny-M 前后向测试，但同一兼容性问题仍存在于 `src/paddlefleet/transformer/moe/moe_layer.py:988`：`fusion_moe_forward` 直接判断 `self.use_accuracy_compatible`。现有未禁用的 `test_latent_moe.py` 以未设置该字段的 `MagicMock` 调用此方法；该属性求值为真，测试会错误进入 `custom_forward` 分支。请同样改为 `getattr(self, "use_accuracy_compatible", False)`，并让现有 latent MoE 用例覆盖默认分支。

File: src/paddlefleet/models/common/language_loss/language_loss.py (hunk 2 above)
Comment (risemeup1111) [reply src/paddlefleet/models/common/language_loss/language_loss.py]: 当前提交已优先读取 `self.pg_collection.ep`，仅在缺失时回退到全局组，并新增显式 EP group 回归测试；该问题已修复。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: src/paddlefleet/transformer/moe/moe_layer.py (hunk 1 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/moe/moe_layer.py]: 当前提交已为 `test_latent_moe.py` 的 `MagicMock` 和 `test_moe_log_balance.py` 的 `SimpleNamespace` 显式设置默认值 `False`；结合此前 `expert_forward` 的 `getattr` 修复及新增回归覆盖，仓库内受影响的直接调用均已具备该属性，该问题已修复。

**Review by risemeup1111 — APPROVED** [review]

**Review by zhengshengning — APPROVED** [review]

**Review by risemeup1111 — APPROVED** [review]
